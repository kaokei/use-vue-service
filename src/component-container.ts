import { ComponentInternalInstance } from 'vue';
import { Container } from '@kaokei/di';

const key = Symbol('container');

export function setContainer(
  component: ComponentInternalInstance,
  container: Container
) {
  (component as any)[key] = container;
}

export function getContainer(component: ComponentInternalInstance) {
  return (component as any)[key];
}
