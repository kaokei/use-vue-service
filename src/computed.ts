import { computed, reactive } from 'vue';

const map = new WeakMap<object, Map<string, any>>();

export function Computed(
  _target: any,
  key: string,
  descriptor: PropertyDescriptor
) {
  const originalGet = descriptor.get!;
  const originalSet = descriptor.set;

  return {
    configurable: true,
    enumerable: true,
    get<T, S = T>() {
      const that = reactive(this);

      if (!map.has(that)) {
        map.set(that, new Map());
      }

      const keyRefMap = map.get(that)!;
      let computedRef = keyRefMap.get(key);

      if (!computedRef) {
        if (originalSet) {
          computedRef = computed({
            get: () => originalGet.call(that) as T,
            set: (v: S) => originalSet.call(that, v),
          });
        } else {
          computedRef = computed<T>(() => originalGet.call(that));
        }
        keyRefMap.set(key, computedRef);
      }

      return computedRef.value;
    },
    set<T>(value: T) {
      const that = reactive(this);
      const computedRef = map.get(that)?.get(key);

      if (computedRef && originalSet) {
        computedRef.value = value;
      }
    },
  } as PropertyDescriptor;
}
