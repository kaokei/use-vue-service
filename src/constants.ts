import { getInjector } from './utils';

export const INJECTOR_KEY = 'USE_VUE_SERVICE_INJECTOR_KEY';
export const DEFAULT_INJECTOR = getInjector();
