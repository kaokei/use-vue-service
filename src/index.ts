export { Inject, Self, Skip, Optional, Injectable } from '@kaokei/di'; // 导出依赖注入需要的装饰器

export { Component } from './Component'; // 支持类组件
export { declareProviders } from './declareProviders'; // setup中声明服务提供者
export { useService } from './useService'; // setup中获取服务实例
