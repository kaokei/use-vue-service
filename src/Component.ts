import { Options } from 'vue-class-component';
import { declareProviders } from './declareProviders';
import { merge } from './utils';
import { INJECTOR_KEY, DEFAULT_INJECTOR } from './constants';
import { injectFromSelf } from './fakeInject';

/**
 * 禁止在组件的构造函数中声明依赖注入
 * 因为类组件需要继承Vue，Vue已经定义了自己的构造函数，导致子类必须在构造函数中手动调用super(props, ctx);
 * 第二点就是组件在实例化的时候，已经写死了`new Ctor(props, ctx);`
 * 没有办法实现依赖注入，假设vue-class-component库支持传递多余的参数`new Ctor(props, ctx, ...rest);`
 * 我认为数据不需要proxy，只需要reactive即可
 *
 * 这里是代理了Options的功能
 *
 * 并且仿造了createDecorator的逻辑来代理setup函数
 *
 * @export
 * @return {*}
 */
export function Component(options: any = {}) {
  return function (target: any) {
    const { providers } = options;
    delete options.providers;

    // 优先执行Options
    // 稍后再补充我们自己的逻辑
    Options(options)(target);

    if (!target.__d) {
      target.__d = [];
    }
    target.__d.push((options: any) => {
      // 注意options.setup是由上方Options方法包装的
      // options本身是没有setup这个属性的
      const originSetup = options.setup;

      options.setup = (props: any, ctx: any) => {
        if (providers) {
          declareProviders(providers);
        }

        const parentInjector = injectFromSelf(INJECTOR_KEY, DEFAULT_INJECTOR);

        // 获取构造函数的实例属性
        // 注意组件不支持构造函数参数注入实例属性
        const properties = parentInjector.getInjectProperties(target);

        // 这里的setupState只能是promise或者plaindata
        // 而不会是render function
        const setupState = originSetup ? originSetup(props, ctx) : {};

        let result: any = null;

        // 这个merge的过程其实最好是写在vue-class-component中
        // 但是估计pr不会被接受
        if (setupState instanceof Promise) {
          result = setupState.then(state => {
            return merge(state, properties);
          });
        } else {
          result = merge(setupState, properties);
        }

        return result;
      };
    });

    return target;
  };
}
