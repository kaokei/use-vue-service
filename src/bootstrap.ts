import { useService } from './useService';
import { DEFAULT_INJECTOR } from './constants';

type UseServiceType = typeof useService;

type DeclareProvidersType = (providers: any[]) => void;

interface InitFn {
  (usFn: UseServiceType, dpFn: DeclareProvidersType): void;
}

export function bootstrap(init: InitFn) {
  const useRootService = (Service: any, options: any) =>
    useService(Service, options, DEFAULT_INJECTOR);

  const declareRootProviders = (providers: any[]) => {
    providers.forEach(provider => DEFAULT_INJECTOR.addProvider(provider));
  };

  init(useRootService, declareRootProviders);
}
