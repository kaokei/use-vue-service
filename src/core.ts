import {
  provide,
  inject,
  getCurrentInstance,
  onUnmounted,
  hasInjectionContext,
  App,
} from 'vue';
import { Container } from '@kaokei/di';
import type { Newable, CommonToken } from '@kaokei/di';
import { createContainer } from './utils';
import { CONTAINER_TOKEN, DEFAULT_CONTAINER } from './constants';

type NewableProvider = Newable[];
type FunctionProvider = (container: Container) => void;
type Provider = NewableProvider | FunctionProvider;

function bindProviders(container: Container, providers: FunctionProvider): void;
function bindProviders(container: Container, providers: NewableProvider): void;
function bindProviders(container: Container, providers: Provider): void;
function bindProviders(container: Container, providers: Provider) {
  if (typeof providers === 'function') {
    providers(container);
  } else {
    for (let i = 0; i < providers.length; i++) {
      const s = providers[i];
      container.bind(s).toSelf();
    }
  }
}

// 查看当前组件是否声明对应的Container
// 参考vue实现
// https://github.com/vuejs/core/blob/733e266cddb945bf9db8edd85a914ecad22a544f/packages/runtime-core/src/apiInject.ts#L47
function getCurrentContainer(): Container | undefined {
  // 注意ComponentInternalInstance类型并没有暴露provides属性，所以这里声明为any类型
  const instance: any = getCurrentInstance();
  if (instance) {
    const token = CONTAINER_TOKEN;
    const provides = instance.provides;
    const parentProvides = instance.parent && instance.parent.provides;
    if (
      // 当前实例的provides存在
      provides &&
      // 当前实例的provides已经和parentProvides不是一个对象，说明已经Object.create(parentProvides)
      provides !== parentProvides &&
      // provides自身已经具有CONTAINER_TOKEN属性
      Object.prototype.hasOwnProperty.call(provides, token)
    ) {
      // 说明当前组件已经声明过对应的Container
      return provides[token];
    }
  }
}
// 从父级组件开始寻找CONTAINER_TOKEN
// 直接借助vue的inject方法
function getProvideContainer(): Container {
  if (hasInjectionContext()) {
    const token = CONTAINER_TOKEN;
    const defaultValue = DEFAULT_CONTAINER;
    return inject(token, defaultValue);
  } else {
    throw new Error('getProvideContainer 只能在 setup 中使用');
  }
}

export function useService<T>(token: CommonToken<T>) {
  const container = getCurrentContainer() || getProvideContainer();
  return container.get(token);
}

export function useRootService<T>(token: CommonToken<T>) {
  return DEFAULT_CONTAINER.get(token);
}

export function useAppService<T>(token: CommonToken<T>, app: App) {
  return app.runWithContext(() => useService(token));
}

export function declareProviders(providers: FunctionProvider): void;
export function declareProviders(providers: NewableProvider): void;
export function declareProviders(providers: Provider) {
  const currentContainer = getCurrentContainer();
  if (currentContainer) {
    bindProviders(currentContainer, providers);
  } else {
    const parent = getProvideContainer();
    const instance = getCurrentInstance();
    let container = createContainer(parent, instance);
    bindProviders(container, providers);
    provide(CONTAINER_TOKEN, container);
    onUnmounted(() => {
      container.unbindAll();
      container = null as any;
    });
  }
}

export function declareRootProviders(providers: FunctionProvider): void;
export function declareRootProviders(providers: NewableProvider): void;
export function declareRootProviders(providers: Provider) {
  bindProviders(DEFAULT_CONTAINER, providers);
}

export function declareAppProviders(
  providers: FunctionProvider,
  app: App
): void;
export function declareAppProviders(providers: NewableProvider, app: App): void;
export function declareAppProviders(providers: Provider, app: App): void;
export function declareAppProviders(providers: Provider, app: App) {
  app.runWithContext(() => {
    const appContainer = inject<Container>(CONTAINER_TOKEN);
    if (appContainer) {
      bindProviders(appContainer, providers);
    } else {
      let container = createContainer(DEFAULT_CONTAINER);
      bindProviders(container, providers);
      app.provide(CONTAINER_TOKEN, container);
      app.onUnmount(() => {
        container.unbindAll();
        container = null as any;
      });
    }
  });
}

export function declareAppProvidersPlugin(providers: Provider) {
  return (app: App) => declareAppProviders(providers, app);
}
