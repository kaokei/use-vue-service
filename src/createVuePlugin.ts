import { INJECTOR_KEY } from './constants';
import { getInjector } from './utils';

export const createVuePlugin = (providers: any = []) => ({
  install: (app: any) => {
    app.provide(INJECTOR_KEY, getInjector(providers));
  },
});
