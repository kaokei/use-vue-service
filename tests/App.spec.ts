import 'reflect-metadata';

import { mount } from '@vue/test-utils';

import TestTreeScope from '@containers/TestTreeScope.vue';

import { DESIGN_PARAM_TYPES, SERVICE_PARAM_TYPES } from '../src';

import Logger from '@services/logger.service';
import Counter from '@services/counter.service';
import Person from '@services/person.service';

describe('App', () => {
  test('渲染组件、获取服务数据', async () => {
    const wrapper = mount(TestTreeScope);
    const Earth = wrapper.findComponent({ name: 'Earth' });
    const China = Earth.findComponent({ name: 'China' });
    const ProvinceA = China.findComponent({ name: 'ProvinceA' });
    const SchoolA = ProvinceA.findComponent({ name: 'SchoolA' });
    const ClassA = SchoolA.findComponent({ name: 'ClassA' });

    const person = (ClassA.vm as any).personService;

    expect(person.check()).toBe(true);

    const types1 = Reflect.getMetadata(DESIGN_PARAM_TYPES, Person);
    const types2 = Reflect.getMetadata(SERVICE_PARAM_TYPES, Person);

    expect(types1[0]).toBe(Logger);
    expect(types2[0]).toBe(Logger);

    expect(types1[1]).toBe(Logger);
    expect(types2[1]).toBe(Logger);

    expect(types1[2]).toBe(Logger);
    expect(types2[2]).toBe(Logger);

    expect(types1[3]).toBe(Counter);
    expect(types2[3]).toBe(Counter);

    expect(types1[4]).toBe(Counter);
    expect(types2[4]).toBe(Counter);

    expect(types1[5]).toBe(Counter);
    expect(types2[5]).toBe(Counter);

    expect(types1[6]).toBe(Counter);
    expect(types2[6]).toBe(Counter);

    expect(types1[7]).toBe(Counter);
    expect(types2[7]).toBe(Counter);

    expect(types1[8]).toBe(Counter);
    expect(types2[8]).toBe(Counter);

    expect(person.logger1).toBeInstanceOf(Logger);
    expect(person.logger2).toBeInstanceOf(Logger);
    expect(person.logger3).toBeInstanceOf(Logger);

    expect(person.counterService11).toBeInstanceOf(Counter);
    expect(person.counterService12).toBeInstanceOf(Counter);

    expect(person.counterService21).toBeInstanceOf(Counter);
    expect(person.counterService22).toBeInstanceOf(Counter);

    expect(person.counterService31).toBeInstanceOf(Counter);
    expect(person.counterService32).toBeInstanceOf(Counter);
  });

  test('组件快照、服务共享', async () => {
    const wrapper = mount(TestTreeScope);
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
