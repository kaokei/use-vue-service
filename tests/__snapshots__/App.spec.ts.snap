// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App 组件快照、服务共享 1`] = `
<div>
  <div>
    <div>
      <div class="container" style="background: rgb(105, 192, 255);"><span class="title">Global Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
    </div>
    <div class="mt5">
      <div class="container" style="background: rgb(105, 192, 255);"><span class="title">Global Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
    </div>
  </div>
  <div>
    <div class="container">
      <h4>Earth</h4>
      <div>
        <div>
          <div class="container" style="background: rgb(105, 192, 255);"><span class="title">Earth Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
        </div>
        <div class="mt5">
          <div class="container" style="background: rgb(92, 219, 211);"><span class="title">Earth Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
        </div>
      </div>
      <div>
        <div class="container">
          <h4>China</h4>
          <div>
            <div>
              <div class="container" style="background: rgb(211, 242, 97);"><span class="title">China Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
            </div>
            <div class="mt5">
              <div class="container" style="background: rgb(92, 219, 211);"><span class="title">China Countdown:</span><button class="decrementBtn" type="button"> - 60</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 60</button></div>
            </div>
          </div>
          <div class="box">
            <div class="container">
              <h4>ProvinceA</h4>
              <div>
                <div>
                  <div class="container" style="background: rgb(255, 192, 105);"><span class="title">ProvinceA Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                </div>
                <div class="mt5">
                  <div class="container" style="background: rgb(92, 219, 211);"><span class="title">ProvinceA Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                </div>
              </div>
              <div class="box">
                <div class="container">
                  <h4>SchoolA</h4>
                  <div>
                    <div>
                      <div class="container" style="background: rgb(255, 156, 110);"><span class="title">SchoolA Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                    </div>
                    <div class="mt5">
                      <div class="container" style="background: rgb(92, 219, 211);"><span class="title">SchoolA Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                    </div>
                  </div>
                  <div class="box">
                    <div class="container">
                      <h4>ClassA</h4>
                      <div>
                        <div>
                          <div class="container" style="background: rgb(255, 120, 117);"><span class="title">ClassA Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                        </div>
                        <div class="mt5">
                          <div class="container" style="background: rgb(255, 156, 110);"><span class="title">SchoolA Counter @Skip:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                        </div>
                        <div class="mt5">
                          <div class="container" style="background: red;" name="Local Person Counter"><span class="title">defaultName:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                        </div>
                        <div class="mt5">
                          <div class="container" style="background: rgb(92, 219, 211);"><span class="title">ClassA Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                        </div>
                      </div>
                    </div>
                    <div class="container">
                      <h4>ClassB</h4>
                      <div>
                        <div>
                          <div class="container" style="background: rgb(255, 156, 110);"><span class="title">ClassB Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                        </div>
                        <div class="mt5">
                          <div class="container" style="background: rgb(92, 219, 211);"><span class="title">ClassB Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="container">
                  <h4>SchoolB</h4>
                  <div>
                    <div>
                      <div class="container" style="background: rgb(255, 192, 105);"><span class="title">SchoolB Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                    </div>
                    <div class="mt5">
                      <div class="container" style="background: rgb(92, 219, 211);"><span class="title">SchoolB Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                    </div>
                  </div>
                  <div class="box">
                    <div class="container">
                      <h4>ClassA</h4>
                      <div>
                        <div>
                          <div class="container" style="background: rgb(255, 192, 105);"><span class="title">ClassA Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                        </div>
                        <div class="mt5">
                          <div class="container" style="background: rgb(92, 219, 211);"><span class="title">ClassA Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                        </div>
                      </div>
                    </div>
                    <div class="container">
                      <h4>ClassB</h4>
                      <div>
                        <div>
                          <div class="container" style="background: rgb(255, 192, 105);"><span class="title">ClassB Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                        </div>
                        <div class="mt5">
                          <div class="container" style="background: rgb(92, 219, 211);"><span class="title">ClassB Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div class="container">
              <h4>ProvinceB</h4>
              <div>
                <div>
                  <div class="container" style="background: rgb(211, 242, 97);"><span class="title">ProvinceB Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                </div>
                <div class="mt5">
                  <div class="container" style="background: rgb(133, 165, 255);"><span class="title">ProvinceB Countdown:</span><button class="decrementBtn" type="button"> - 30</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 30</button></div>
                </div>
              </div>
              <div class="box">
                <div class="container">
                  <h4>SchoolA</h4>
                  <div>
                    <div>
                      <div class="container" style="background: rgb(211, 242, 97);"><span class="title">SchoolA Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                    </div>
                    <div class="mt5">
                      <div class="container" style="background: rgb(133, 165, 255);"><span class="title">SchoolA Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                    </div>
                  </div>
                  <div class="box">
                    <div class="container">
                      <h4>ClassA</h4>
                      <div>
                        <div>
                          <div class="container" style="background: rgb(211, 242, 97);"><span class="title">ClassA Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                        </div>
                        <div class="mt5">
                          <div class="container" style="background: rgb(133, 165, 255);"><span class="title">ClassA Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                        </div>
                      </div>
                    </div>
                    <div class="container">
                      <h4>ClassB</h4>
                      <div>
                        <div>
                          <div class="container" style="background: rgb(211, 242, 97);"><span class="title">ClassB Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                        </div>
                        <div class="mt5">
                          <div class="container" style="background: rgb(133, 165, 255);"><span class="title">ClassB Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="container">
                  <h4>SchoolB</h4>
                  <div>
                    <div>
                      <div class="container" style="background: rgb(211, 242, 97);"><span class="title">SchoolB Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                    </div>
                    <div class="mt5">
                      <div class="container" style="background: rgb(179, 127, 235);"><span class="title">SchoolB Countdown:</span><button class="decrementBtn" type="button"> - 20</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 20</button></div>
                    </div>
                  </div>
                  <div class="box">
                    <div class="container">
                      <h4>ClassA</h4>
                      <div>
                        <div>
                          <div class="container" style="background: rgb(211, 242, 97);"><span class="title">ClassA Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                        </div>
                        <div class="mt5">
                          <div class="container" style="background: rgb(179, 127, 235);"><span class="title">ClassA Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                        </div>
                      </div>
                    </div>
                    <div class="container">
                      <h4>ClassB</h4>
                      <div>
                        <div>
                          <div class="container" style="background: rgb(211, 242, 97);"><span class="title">ClassB Counter:</span><button class="decrementBtn" type="button"> - </button><span class="countNum">0</span><button class="incrementBtn" type="button"> + </button></div>
                        </div>
                        <div class="mt5">
                          <div class="container" style="background: rgb(255, 133, 192);"><span class="title">ClassB Countdown:</span><button class="decrementBtn" type="button"> - 10</button><span class="countNum">00:00:00</span><button class="incrementBtn" type="button"> + 10</button></div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="switch-box">
    <h4> These areas with same background color have same service instance. </h4>
    <div><button type="button">hide counter background </button></div>
    <div class="mt5"><button type="button">hide countdown background </button></div>
  </div>
</div>
`;
